synapse/push/pusherpool.py

```diff
@@ -330,9 +343,11 @@ def remove_pusher(self, app_id, pushkey, user_id):
 
         if appid_pushkey in byuser:
             logger.info("Stopping pusher %s / %s", user_id, appid_pushkey)
-            byuser[appid_pushkey].on_stop()
-            with self._pushers_lock:
-                del byuser[appid_pushkey]
+            pusher = byuser[appid_pushkey]
+            pusher.on_stop()
+            del byuser[appid_pushkey]
```

Comment:
I think you can `pop()` here, should avoid looking it up in the hash a second time.